core/iam_authenticator.go

```diff
@@ -167,24 +170,74 @@ func (this IamAuthenticator) Validate() error {
 }
 
 // getToken: returns an access token to be used in an Authorization header.
-// Whenever a new token is needed (when a token doesn't yet exist, or the existing token has expired),
-// a new access token is fetched from the token server.
+// Whenever a new token is needed (when a token doesn't yet exist, needs to be refreshed,
+// or the existing token has expired), a new access token is fetched from the token server.
 func (authenticator *IamAuthenticator) getToken() (string, error) {
 	if authenticator.tokenData == nil || !authenticator.tokenData.isTokenValid() {
-		tokenResponse, err := authenticator.requestToken()
+		// synchronously request the token
+		err := authenticator.synchronizedRequestToken()
 		if err != nil {
 			return "", err
 		}
-
-		authenticator.tokenData, err = newIamTokenData(tokenResponse)
-		if err != nil {
-			return "", err
+	} else if authenticator.tokenData.needsRefresh() {
+		// If refresh needed, kick off a go routine in the background to get a new token
+		ch := make(chan error)
+		go func() {
+			ch <- authenticator.getTokenData()
+		}()
+		select {
+		case err := <-ch:
+			if err != nil {
+				return "", err
+			}
+		default:
 		}
 	}
 
+	// return an error if the access token is not valid or was not fetched
+	if authenticator.tokenData == nil || authenticator.tokenData.AccessToken == "" {
```

Comment:
It's curious that here you check `authenticator.tokenData.AccessToken == ""` but above you check `!authenticator.tokenData.isTokenValid()`.  Why the difference?  Can't we make them both the same?

Reply:
Just an oversight on my end. I'll go ahead and add that

Reply:
So @mkistler I went ahead and tried to make this change, though it's important to note``isTokenValid`` also checks to see if the token is expired. I think it's a good extra check to have, though the issue is that for cp4d's unit tests, the expiration time is calculated using some jwt token, and not by a mock value from actual token response. The token used in those tests is expired, so all the tests using it would fail.. Getting the unit tests to work would require either a new token that would eventually also expire, changing the ``GetCurrentTime`` method to accept a mock clock, or changing how we set and calculate the expired time for a token. I just went ahead and reverted this change.